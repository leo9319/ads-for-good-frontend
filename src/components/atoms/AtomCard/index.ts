export { default } from './AtomCard';
export { AtomCard } from './AtomCard';
export type { AtomCardProps } from './AtomCard';
