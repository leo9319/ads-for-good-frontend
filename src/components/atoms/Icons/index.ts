// export { IconDisplay } from './Icons'
export { default } from './Icon';
export { Icon } from './Icon';
