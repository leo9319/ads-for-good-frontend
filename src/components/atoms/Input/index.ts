export type { InputProps } from './Input';
export { default } from './Input';
export { Input } from './Input';
