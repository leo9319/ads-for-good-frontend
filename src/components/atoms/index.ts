export * from './VideoPlayer';
export * from './TextCaption';
export * from './Icons';
export * from './Input';
export * from './Image';
export * from './Truncate';
export * from './AtomCard';
