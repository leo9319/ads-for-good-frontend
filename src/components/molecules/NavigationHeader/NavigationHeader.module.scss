@import '../../../styles/breakpoints';
@import '../../../styles/colours';

.headerWrapper {
  min-height: 56px;
  padding: 0px 60px;
  box-sizing: border-box;
  background-color: $color-white;
  box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.1);

  .headerInner {
    max-width: 1319.6px;
    margin: 0 auto;
    display: flex;
    align-items: center;

    .hamburgerMenu {
      display: none;
    }

    .logoSection {
      width: 20%;
      text-align: center;

      .logo {
        width: 159.5px !important;
        height: 33px !important;

        &.clickable {
          cursor: pointer;
        }
      }
    }

    .navCtaSection {
      width: 40%;
      display: flex;
      justify-content: end;

      .navCtaSectionInner {
        button {
          font-size: var(--spacing-md);
          padding: 11.67px 20.67px;
        }
      }
    }
  }
}

.navLinksSection {
  width: 40%;

  ul {
    list-style: none;
    margin: 0;
    padding: 0;
    display: flex;

    li {
      a {
        float: left;
        padding: 29px var(--spacing-md);
        box-sizing: border-box;
        font-size: var(--spacing-md);
        font-weight: 600;
        color: $color-neutral-600;
        cursor: pointer;
        text-decoration: none;
        border-bottom: 2px solid transparent;
        box-shadow: none !important;
        outline: none !important;

        &:hover {
          color: $color-neutral-950;
        }

        &:focus {
          box-shadow: none;
        }
      }

      &.activeNavLink {
        a {
          color: $color-neutral-950;
          border-color: $color-orange-600;
        }
      }
    }
  }

  &.fullWidth {
    width: 100%;

    ul li a {
      padding: 19px var(--spacing-md);
    }
  }

  &.tabCentered ul {
    justify-content: center;
  }
}

@media (max-width: $sm-min) {
  .headerWrapper {
    padding: 0 var(--spacing-2xl);

    .headerInner {
      justify-content: left;
      min-height: 56px;

      .navLinksSection {
        width: 100%;
        position: fixed;
        z-index: 4000;
        left: -100%;
        opacity: 0;
        top: 59px;
        height: 100%;
        background-color: $color-white;
        transition: all 0.3s ease;

        &.navLinksSlideIn {
          left: 0;
          opacity: 1;
          transition: all 0.3s ease;
        }

        ul {
          flex-direction: column;
          overflow: auto;
          height: calc(100vh - 170px);
          padding-top: var(--spacing-sm);

          li {
            a {
              border: none;
              padding: 12px var(--spacing-2xl);
            }
          }
        }
      }

      .hamburgerMenu {
        display: block;
        width: 38%;

        .hamburgerMenuIcon {
          justify-content: left !important;
          width: 24px;

          span {
            cursor: pointer;
          }
        }
      }

      .logoSection {
        .logo {
          width: 116px !important;
          height: 24px !important;
        }
      }

      .navCtaSection {
        justify-content: left;
        position: fixed;
        z-index: 4000;
        bottom: 0px;
        width: 100%;
        left: -100%;
        opacity: 0;
        padding: 24px var(--spacing-2xl);
        box-sizing: border-box;
        transition: all 0.3s ease;
        background-color: $color-white;
        border-top: 1px solid $color-neutral-300;

        &.navLinksSlideIn {
          left: 0;
          opacity: 1;
          transition: all 0.3s ease;
        }

        .navCtaSectionInner {
          width: 100%;

          button {
            width: 100%;
          }
        }
      }
    }
  }
}

@media (max-width: $xs-min) {
  .headerWrapper {
    padding: 0 var(--spacing-md);

    .headerInner {
      .navLinksSection {
        top: 56px;

        ul {
          li {
            a {
              padding: 12px var(--spacing-md);
            }
          }
        }
      }

      .hamburgerMenu {
        width: 35%;
      }

      .navCtaSection {
        padding: 24px var(--spacing-md);
      }
    }
  }
}
