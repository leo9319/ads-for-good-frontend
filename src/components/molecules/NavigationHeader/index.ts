export { default } from './NavigationHeader';
export { NavigationHeader } from './NavigationHeader';
export { Tabs } from './NavigationHeader';
export type { NavigationHeaderProps, NavigationData } from './NavigationHeader';
