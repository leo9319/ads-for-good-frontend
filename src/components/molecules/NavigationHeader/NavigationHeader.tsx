import React, { useEffect, useId, useState } from 'react';
import Image from '@components/atoms/Image';
import { classNames, getModuleClasses } from '@utils/common/classNames';
import FalbackImage from '@assets/images/FallbackImage.svg';
import WVLogoDesktop from '@assets/icons/logo/LogoDark.svg';
import ButtonGroup from '@radix-styles/molecules/ButtonGroup';
import { Icon } from '@components/atoms/Icons';
import { ButtonProps } from '@radix-styles/atoms/Button';
import Link from '@radix-styles/atoms/Link';

import styles from './NavigationHeader.module.scss';

type NavigationItem = {
  /**
   * Name of the nav link.
   * @property
   */
  name: string;
  /**
   * Fragment Id for the nav.
   * @property
   */
  fragmentId: string;
  /**
   * Link for the nav.
   * @property
   */
  path: string;
  /**
   * Nav navigable.
   * @property
   */
  navigable: boolean;
  /**
   * Child menu fro the nav.
   * @property
   */
  children: NavigationItem[];
};

export type NavigationData = {
  navigation: NavigationItem[];
};

export interface NavigationHeaderProps {
  /**
   * Optional class name for the header component.
   * @property
   */
  className?: string;
  /**
   * Optional, header logo alt text.
   * @property
   * @default 'logo'
   */
  logoAltText?: string;
  /**
   * Navigation links
   * @property
   */
  navLinks?: NavigationData;
  /**
   * List of buttons to display in the group
   * @property
   */
  buttonData?: Array<ButtonProps>;
  /**
   * Optional on click handler for the logo.
   * @property
   * @returns void
   */
  onLogoClick?: () => void;
  /**
   * Optional on click handler for the Nav Links.
   * @property
   * @returns void
   */
  onNavLinkClick?: (
    path: string,
    event: React.MouseEvent<HTMLAnchorElement>
  ) => void;
  /**
   * Get the active path to highlight menu item
   * @property
   */
  activePath?: string;
}

export const Tabs = ({
  navLinks,
  menuOpen,
  keyPrefix = '',
  onLinkClick,
  fullWidth = false,
  isCentered = false,
  activePath,
}: {
  navLinks?: NavigationData;
  menuOpen?: boolean;
  keyPrefix?: string;
  onLinkClick: (
    path: string,
    event: React.MouseEvent<HTMLAnchorElement>
  ) => void;
  fullWidth?: boolean;
  isCentered?: boolean;
  activePath?: string;
}) => {
  const [activeIndex, setActiveIndex] = useState<unknown>(null);
  const onNavClick = (
    path: string,
    event: React.MouseEvent<HTMLAnchorElement>,
    index: number | null
  ) => {
    event.preventDefault();
    setActiveIndex(index);
    onLinkClick(path, event);
  };

  useEffect(() => {
    if (typeof window !== 'undefined' && navLinks?.navigation && activePath) {
      const currentIndex = navLinks.navigation.findIndex(value =>
        activePath.includes(value.path)
      );
      setActiveIndex(currentIndex > 0 ? currentIndex : 0);
    }
  }, [activePath]);

  return (
    <nav
      className={classNames(
        styles.navLinksSection,
        menuOpen ? styles.navLinksSlideIn : '',
        fullWidth ? styles.fullWidth : '',
        isCentered ? styles.tabCentered : ''
      )}
    >
      <ul role="menu" aria-label="Main menu">
        {navLinks?.navigation?.map((value, index) => (
          <li
            key={`${keyPrefix}-${index}`}
            className={classNames(
              activeIndex === index ? styles.activeNavLink : ''
            )}
            role="menuitem"
          >
            <Link
              href={value.path}
              onClick={event => onNavClick(value.path, event, index)}
            >
              {value.name}
            </Link>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export const NavigationHeader = ({
  className,
  logoAltText = 'logo',
  navLinks,
  buttonData = [],
  onLogoClick,
  onNavLinkClick,
  activePath = '',
}: NavigationHeaderProps) => {
  const classes = getModuleClasses(className?.trim(), styles);
  const key = useId();
  const [menuOpen, setMenuOpen] = useState(false);

  useEffect(() => {
    const handleResize = () => {
      if (window.innerWidth <= 768) {
        setMenuOpen(false);
      }
    };
    window.addEventListener('resize', handleResize);
    handleResize();
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  useEffect(() => {
    if (menuOpen) {
      const scrollBarWidth =
        window.innerWidth - document.documentElement.clientWidth;
      document.body.style.overflow = 'hidden';
      document.body.style.paddingRight = `${scrollBarWidth}px`;
    }

    return () => {
      document.body.style.overflow = '';
      document.body.style.paddingRight = '';
    };
  }, [menuOpen]);

  const onLinkClick = (
    path: string,
    event: React.MouseEvent<HTMLAnchorElement>
  ) => {
    onNavLinkClick?.(path, event);
    setMenuOpen(false);
  };

  return (
    <div className={classNames(styles.headerWrapper, classes)}>
      <div className={classNames(styles.headerInner)}>
        <div className={classNames(styles.hamburgerMenu)}>
          <Icon
            className={classNames(styles.hamburgerMenuIcon)}
            name={menuOpen ? 'icon-close' : 'icon-menu'}
            size="24"
            onClick={() => setMenuOpen(!menuOpen)}
          />
        </div>

        <Tabs
          navLinks={navLinks}
          menuOpen={menuOpen}
          keyPrefix={key}
          onLinkClick={onLinkClick}
          activePath={activePath}
        />

        <div className={classNames(styles.logoSection)}>
          <Image
            onClick={onLogoClick}
            src={WVLogoDesktop}
            alt={logoAltText}
            fallbackSrc={FalbackImage}
            className={classNames(
              styles.logo,
              onLogoClick ? styles.clickable : ''
            )}
          />
        </div>

        {buttonData.length > 0 ? (
          <div
            className={classNames(
              styles.navCtaSection,
              menuOpen ? styles.navLinksSlideIn : ''
            )}
          >
            <div className={classNames(styles.navCtaSectionInner)}>
              <ButtonGroup data={buttonData} />
            </div>
          </div>
        ) : null}
      </div>
    </div>
  );
};

export default NavigationHeader;
