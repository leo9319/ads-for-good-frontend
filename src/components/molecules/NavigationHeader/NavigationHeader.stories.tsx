import type {
  Decorator,
  Meta,
  ReactRenderer,
  StoryObj,
  StrictArgs,
} from '@storybook/react';
import NavigationHeader, { NavigationData } from './NavigationHeader';
import { ButtonProps } from '@radix-styles/atoms/Button';
import {
  Route,
  BrowserRouter as Router,
  Routes,
  useNavigate,
} from 'react-router-dom';
import React, { useEffect, useState } from 'react';
import { PartialStoryFn } from 'storybook/internal/types';

const navigationData: NavigationData = {
  navigation: [
    {
      name: 'Home',
      fragmentId: '/en/home',
      path: '/en/home',
      navigable: true,
      children: [],
    },
    {
      name: 'Learn',
      fragmentId:
        '/content/dam/wvc/EN/context-specific-content/one-off-pages/learn/body/container',
      path: '/en/learn',
      navigable: true,
      children: [],
    },
    {
      name: 'Stories',
      fragmentId: 'en/stories',
      path: '/en/stories',
      navigable: true,
      children: [],
    },
  ],
};

const buttonData: Array<ButtonProps> = [
  {
    text: 'Sign Up',
    mode: 'primary',
    size: 'sm',
    onClick: () => {
      console.log('Button clicked');
    },
  },
];

const RouterProvider: Decorator = Story => {
  const [key, setKey] = useState(0);
  const style = { margin: '10px' };

  useEffect(() => {
    setKey(prevKey => prevKey + 1);
  }, [Story]);

  return (
    <Router key={key}>
      <RouterContent Story={Story} style={style} />
    </Router>
  );
};

const RouterContent = ({
  Story,
  style,
}: {
  Story: PartialStoryFn<ReactRenderer, StrictArgs>;
  style: React.CSSProperties;
}) => {
  const navigate = useNavigate();

  const onNavLinkClick = (path: string) => {
    console.log(`Navigating to: ${path}`);
    navigate(path);
  };

  const onLogoClick = () => {
    navigate(navigationData.navigation[0].path);
  };

  return (
    <>
      <Story onNavLinkClick={onNavLinkClick} onLogoClick={onLogoClick} />
      <Routes>
        <Route
          path={navigationData.navigation[0].path}
          element={<div style={style}>{navigationData.navigation[0].name}</div>}
        />
        <Route
          path={navigationData.navigation[1].path}
          element={<div style={style}>{navigationData.navigation[1].name}</div>}
        />
        <Route
          path={navigationData.navigation[2].path}
          element={<div style={style}>{navigationData.navigation[2].name}</div>}
        />
      </Routes>
    </>
  );
};

const meta = {
  title: 'Library/Molecules/NavigationHeader',
  component: NavigationHeader,
  parameters: {
    layout: 'fullscreen',
  },
  tags: ['autodocs'],
  argTypes: {
    className: { control: { type: 'text' } },
    logoAltText: { control: { type: 'text' } },
  },
  args: {
    className: '',
    logoAltText: 'logo',
    navLinks: navigationData,
    buttonData: buttonData,
  },
  decorators: [RouterProvider],
} satisfies Meta<typeof NavigationHeader>;

export default meta;
type Story = StoryObj<typeof meta>;

export const NavigationHeaderView: Story = {
  args: {},
};
