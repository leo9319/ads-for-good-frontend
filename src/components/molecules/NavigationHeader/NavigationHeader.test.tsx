import React from 'react';
import { render } from '@testing-library/react';
import NavigationHeader, { NavigationData } from './NavigationHeader';
import { ButtonProps } from '@radix-styles/atoms/Button';

const navigationData: NavigationData = {
  navigation: [
    {
      name: 'Home',
      fragmentId: '/en/home',
      path: '/en/home',
      navigable: true,
      children: [],
    },
    {
      name: 'Ways to Give',
      fragmentId: 'wvc/ways-to-give',
      path: '/en/give',
      navigable: true,
      children: [],
    },
    {
      name: 'Learn',
      fragmentId:
        '/content/dam/wvc/EN/context-specific-content/one-off-pages/learn/body/container',
      path: '/en/learn',
      navigable: true,
      children: [],
    },
  ],
};

const buttonData: Array<ButtonProps> = [
  {
    text: 'Sign Up',
    mode: 'primary',
    size: 'sm',
    onClick: () => {
      console.log('Button clicked');
    },
  },
];

describe('Navigation Header', () => {
  it('Should render successfully', () => {
    const { baseElement } = render(<NavigationHeader />);
    expect(baseElement).toMatchSnapshot();
  });
  it('Should render successfully with nav data', () => {
    const { baseElement } = render(
      <NavigationHeader navLinks={navigationData} />
    );
    expect(baseElement).toMatchSnapshot();
  });
  it('Should render successfully with nav data and button data', () => {
    const { baseElement } = render(
      <NavigationHeader navLinks={navigationData} buttonData={buttonData} />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
