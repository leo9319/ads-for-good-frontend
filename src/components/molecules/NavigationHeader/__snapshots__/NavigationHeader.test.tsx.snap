// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Header Should render successfully 1`] = `
<body>
  <div>
    <div
      class="headerWrapper"
    >
      <div
        class="headerInner"
      >
        <div
          class="hamburgerMenu"
        >
          <span
            class="hamburgerMenuIcon"
            style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
          >
            <span
              class="icon icon-menu"
              style="width: 24px; height: 24px;"
            />
          </span>
        </div>
        <nav
          class="navLinksSection"
        >
          <ul
            aria-label="Main menu"
            role="menu"
          />
        </nav>
        <div
          class="logoSection"
        >
          <img
            alt="logo"
            aria-hidden="true"
            class="rt-Skeleton imgWrapper logo"
            inert=""
            src="SvgrURL"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Navigation Header Should render successfully with nav data 1`] = `
<body>
  <div>
    <div
      class="headerWrapper"
    >
      <div
        class="headerInner"
      >
        <div
          class="hamburgerMenu"
        >
          <span
            class="hamburgerMenuIcon"
            style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
          >
            <span
              class="icon icon-menu"
              style="width: 24px; height: 24px;"
            />
          </span>
        </div>
        <nav
          class="navLinksSection"
        >
          <ul
            aria-label="Main menu"
            role="menu"
          >
            <li
              class="activeNavLink"
              role="menuitem"
            >
              <a
                class="rt-Text rt-reset rt-Link rt-underline-auto link body-md"
                data-accent-color=""
                href="/en/home"
                tabindex="0"
              >
                Home
              </a>
            </li>
            <li
              class=""
              role="menuitem"
            >
              <a
                class="rt-Text rt-reset rt-Link rt-underline-auto link body-md"
                data-accent-color=""
                href="/en/give"
                tabindex="0"
              >
                Ways to Give
              </a>
            </li>
            <li
              class=""
              role="menuitem"
            >
              <a
                class="rt-Text rt-reset rt-Link rt-underline-auto link body-md"
                data-accent-color=""
                href="/en/learn"
                tabindex="0"
              >
                Learn
              </a>
            </li>
          </ul>
        </nav>
        <div
          class="logoSection"
        >
          <img
            alt="logo"
            aria-hidden="true"
            class="rt-Skeleton imgWrapper logo"
            inert=""
            src="SvgrURL"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Navigation Header Should render successfully with nav data and button data 1`] = `
<body>
  <div>
    <div
      class="headerWrapper"
    >
      <div
        class="headerInner"
      >
        <div
          class="hamburgerMenu"
        >
          <span
            class="hamburgerMenuIcon"
            style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
          >
            <span
              class="icon icon-menu"
              style="width: 24px; height: 24px;"
            />
          </span>
        </div>
        <nav
          class="navLinksSection"
        >
          <ul
            aria-label="Main menu"
            role="menu"
          >
            <li
              class="activeNavLink"
              role="menuitem"
            >
              <a
                class="rt-Text rt-reset rt-Link rt-underline-auto link body-md"
                data-accent-color=""
                href="/en/home"
                tabindex="0"
              >
                Home
              </a>
            </li>
            <li
              class=""
              role="menuitem"
            >
              <a
                class="rt-Text rt-reset rt-Link rt-underline-auto link body-md"
                data-accent-color=""
                href="/en/give"
                tabindex="0"
              >
                Ways to Give
              </a>
            </li>
            <li
              class=""
              role="menuitem"
            >
              <a
                class="rt-Text rt-reset rt-Link rt-underline-auto link body-md"
                data-accent-color=""
                href="/en/learn"
                tabindex="0"
              >
                Learn
              </a>
            </li>
          </ul>
        </nav>
        <div
          class="logoSection"
        >
          <img
            alt="logo"
            aria-hidden="true"
            class="rt-Skeleton imgWrapper logo"
            inert=""
            src="SvgrURL"
            tabindex="-1"
          />
        </div>
        <div
          class="navCtaSection"
        >
          <div
            class="navCtaSectionInner"
          >
            <div
              class="rt-Flex rt-r-display-flex rt-r-fd-row rt-r-gap-3 buttonGroupContainer"
            >
              <button
                aria-label="Sign Up"
                class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton sm primary none rt-Box buttonWrapper"
                data-accent-color=""
              >
                <div
                  class="buttonContent textOnly"
                >
                  <span>
                    Sign Up
                  </span>
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
